mlflow2seldon/api.py

```diff
@@ -0,0 +1,291 @@
+                            image=deploy_image_ref,
+                            model_storage_uri=storage_art_uri,
+                            model_name=model.name,
+                            model_stage=model_version.current_stage,
+                            model_version=model_version.version,
+                            source_run_id=model_version.run_id,
+                            deployment_namespace=seldon_deployment_ns,
+                        )
+                        deployed_model = seldon_models.get(registered_model.name)
+                        if not deployed_model or not deployed_model.is_same_version(
+                            registered_model
+                        ):
+                            registered_model.need_redeploy = True
+                        else:
+                            registered_model.need_redeploy = False
+                        mlflow_models[registered_model.name] = registered_model
+
+                # deploy models in Seldon
+                for model_name in mlflow_models.keys():
+                    if mlflow_models[model_name].need_redeploy:
+                        await _deploy_model(
+                            mlflow_models[model_name],
+                            neuro_client,
+                            registry_secret_name,
+                        )
+                        mlflow_models[model_name].need_redeploy = False
+
+                # removing outdated models
+                for model_name in set(seldon_models.keys()) - set(mlflow_models.keys()):
+                    _delete_seldon_deployment(seldon_models[model_name])
+
+                # the state is sync
+                seldon_models = mlflow_models.copy()
+                deployed_models = [
+                    f"{n}:{m.model_version}" for n, m in seldon_models.items()
+                ]
+                logging.info(f"Deployed models: {';'.join(deployed_models)}")
+                await asyncio.sleep(DELAY)  # not to overload an API
+
+            except KeyboardInterrupt:
+                logging.warning("Got keyboard interrupt, gracefully shutting down...")
+                break
+            except MlflowException as e:
+                if "Page Not Found" in e.message:
+                    logging.warning(
+                        "Unable to connect to the MLFlow server. "
+                        "Is it running as a platform job?"
+                    )
+                elif "Sign In" in e.message:
+                    logging.warning(
+                        "Unable to connect to the MLFlow server - auth is required. "
+                        "SSO is disabled?"
+                    )
+                else:
+                    logging.warning(f"MLFlow server error occured:\n{e}")
+            except Exception as e:
+                logging.warning(f"Unexpected exception (ignoring): {e}")
+    finally:
+        for model in seldon_models.values():
+            _delete_seldon_deployment(model)
```

Comment:
Hmm not sure if we want to delete seldon deployments once the service if terminated. For demo -- OK, but for prod it's very wrong. Let's add an explicit flag "cleanup deployed models on finish"?

Reply:
because we might get into the finally branch even if the pod gets killed or evicted (not sure -- it'll be SIGTERM then graceful period ~30 seconds, then SIGKILL).

Well, alternatively we could just drop the outer try-catch